### PR TITLE
core: clear temporary stack flag before entering boot_init_primary_late()

### PR DESCRIPTION
**Pull request by jforissier**
boot_init_primary_late() uses the stack of thread 0, so the flag that
indicates usage of the temporary stack must be cleared in the current
core's thread_core_local structure.

Fixes the following crash when CFG_CORE_DEBUG_CHECK_STACKS=y:

 D/TC:0 0 select_vector:1126 SMCCC_ARCH_WORKAROUND_1 (0x80008000) available
 D/TC:0 0 select_vector:1128 SMC Workaround for CVE-2017-5715 used
 D/TC:0 0 check_stack_limits:370 Stack pointer out of range (0xb7f54fd0)
 D/TC:0 0 print_stack_limits:346 tmp [0] 0xb7f57c90..0xb7f584b0
 D/TC:0 0 print_stack_limits:346 tmp [1] 0xb7f58ad0..0xb7f592f0
 D/TC:0 0 print_stack_limits:346 tmp [2] 0xb7f59910..0xb7f5a130
 D/TC:0 0 print_stack_limits:346 tmp [3] 0xb7f5a750..0xb7f5af70
 D/TC:0 0 print_stack_limits:351 abt [0] 0xb7f4e710..0xb7f4f330
 D/TC:0 0 print_stack_limits:351 abt [1] 0xb7f4f950..0xb7f50570
 D/TC:0 0 print_stack_limits:351 abt [2] 0xb7f50b90..0xb7f517b0
 D/TC:0 0 print_stack_limits:351 abt [3] 0xb7f51dd0..0xb7f529f0
 D/TC:0 0 print_stack_limits:356 thr [0] 0xb7f53030..0xb7f55030
 D/TC:0 0 print_stack_limits:356 thr [1] 0xb7f55670..0xb7f57670
 E/TC:0 0 Panic at core/arch/arm/kernel/thread.c:372 <check_stack_limits>

Signed-off-by: Jerome Forissier <jerome@forissier.org>

### CODE REVIEW

File: core/arch/arm/kernel/entry_a32.S [hunk 1]
```diff
@@ -551,10 +553,19 @@ shadow_stack_access_ok:
 	ldr	r0, =threads
 	ldr	r0, [r0, #THREAD_CTX_STACK_VA_END]
 	mov	sp, r0
+	bl	__get_core_pos
+	mov	r1, #THREAD_CORE_LOCAL_SIZE
+	ldr	r2, =thread_core_local
+	mla	r8, r0, r1, r2
+	ldr	r10, [r8, #THREAD_CORE_LOCAL_FLAGS]
+	bic	r10, r10, #THREAD_CLF_TMP
+	str	r10, [r8, #THREAD_CORE_LOCAL_FLAGS]
```
Comment (jforissier) [inline core/arch/arm/kernel/entry_a32.S:562]: Or perhaps:
```suggestion
	mov	r10, #0
	str	r10, [r8, #THREAD_CORE_LOCAL_FLAGS]
```
At this point it seems `flags` can only be `== THREAD_CLF_TMP` with no other bit set, or am I missing something?

File: core/arch/arm/kernel/entry_a32.S [hunk 2]
```diff
@@ -551,10 +553,19 @@ shadow_stack_access_ok:
 	ldr	r0, =threads
 	ldr	r0, [r0, #THREAD_CTX_STACK_VA_END]
 	mov	sp, r0
+	bl	__get_core_pos
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/entry_a32.S:556]: Would it make sense to call `thread_get_core_local()` instead?

File: core/arch/arm/kernel/entry_a32.S (hunk 1 above)
Comment (jenswi-linaro) [reply core/arch/arm/kernel/entry_a32.S:562]: Agree

File: core/arch/arm/kernel/entry_a32.S (hunk 2 above)
Comment (jforissier) [reply core/arch/arm/kernel/entry_a32.S]: Totally. Fixed in new version.

File: core/arch/arm/kernel/entry_a32.S (hunk 1 above)
Comment (jforissier) [reply core/arch/arm/kernel/entry_a32.S]: Fixed

**Comment by jforissier** [comment]
Update

File: core/arch/arm/kernel/asm-defines.c [hunk 3]
```diff
@@ -95,6 +95,7 @@ DEFINES
 	/* struct thread_core_local */
 	DEFINE(THREAD_CORE_LOCAL_X0, offsetof(struct thread_core_local, x[0]));
 	DEFINE(THREAD_CORE_LOCAL_X2, offsetof(struct thread_core_local, x[2]));
+	DEFINE(THREAD_CORE_LOCAL_SIZE, sizeof(struct thread_core_local));
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/asm-defines.c:98]: This isn't needed any longer.

**Review by jenswi-linaro — COMMENTED** [review]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/arch/arm/kernel/asm-defines.c (hunk 3 above)
Comment (jforissier) [reply core/arch/arm/kernel/asm-defines.c]: Fixed

**Pull request merged**